File: Windows/CWindow.vb
Public MustInherit Class CWindow(Of T As {CWindowData, New})
    Inherits Window

    Public ReadOnly Property WindowData As New T

    Public Sub New()
        WindowState = WindowState.Normal
        WindowStartupLocation = WindowStartupLocation.CenterScreen
    End Sub

    Private Sub Window_Loaded(sender As Object, e As EventArgs) Handles Me.Loaded
        Me.DataContext = WindowData
        WindowData.InitWindow(Me)
    End Sub
End Class

File: Windows/CWindow.cs

using System.Windows;

public abstract class CWindow<T> : Window where T : CWindowData, new()
{
    public T WindowData { get; private set; } = new T();

    public CWindow()
    {
        WindowStartupLocation = WindowStartupLocation.CenterScreen;

        Loaded += (s, e) =>
        {
            DataContext = WindowData;
            WindowData.InitWindow(this);
        };
    }
}